virtual_ship/instruments/argo.py

```diff
@@ -0,0 +1,168 @@
+        particle_ddepth -= particle.vertical_speed * particle.dt
+        particle.cycle_age += (
+            particle.dt
+        )  # solve issue of not updating cycle_age during ascent
+        if particle.depth + particle_ddepth >= particle.min_depth:
+            particle_ddepth = particle.min_depth - particle.depth
+            particle.temperature = (
+                math.nan
+            )  # reset temperature to NaN at end of sampling cycle
+            particle.salinity = math.nan  # idem
+            particle.cycle_phase = 4
+        else:
+            particle.temperature = fieldset.T[
+                time, particle.depth, particle.lat, particle.lon
+            ]
+            particle.salinity = fieldset.S[
+                time, particle.depth, particle.lat, particle.lon
+            ]
+
+    elif particle.cycle_phase == 4:
+        # Phase 4: Transmitting at surface until cycletime is reached
+        if particle.cycle_age > particle.cycle_days * 86400:
+            particle.cycle_phase = 0
+            particle.cycle_age = 0
+
+    if particle.state == StatusCode.Evaluate:
+        particle.cycle_age += particle.dt  # update cycle_age
+
+
+def _keep_at_surface(particle, fieldset, time):
+    # Prevent error when float reaches surface
+    if particle.state == StatusCode.ErrorThroughSurface:
+        particle.depth = particle.min_depth
+        particle.state = StatusCode.Success
+
+
+def _check_error(particle, fieldset, time):
+    if particle.state >= 50:  # This captures all Errors
+        particle.delete()
+
+
+def simulate_argos(
+    argos: list[Argo],
+    fieldset: FieldSet,
+    out_file_name: str,
+) -> None:
+    """
+    Use parcels to simulate a set of argos in a fieldset.
+
+    :param argos: The argos to simulate.
+    :param fieldset: The fieldset to simulate the argos in.
+    :param out_file_name: The file to write the results to.
+    """
+    lon = [argo.location.lon for argo in argos]
+    lat = [argo.location.lat for argo in argos]
+    time = [argo.deployment_time for argo in argos]
+
+    # define the parcels simulation
+    argoset = ParticleSet(
+        fieldset=fieldset,
+        pclass=_ArgoParticle,
+        lon=lon,
+        lat=lat,
+        depth=[argo.min_depth for argo in argos],
+        time=time,
+        min_depth=[argo.min_depth for argo in argos],
+        max_depth=[argo.max_depth for argo in argos],
+        drift_depth=[argo.drift_depth for argo in argos],
+        vertical_speed=[argo.vertical_speed for argo in argos],
+        cycle_days=[argo.cycle_days for argo in argos],
+        drift_days=[argo.drift_days for argo in argos],
+    )
+
+    # define output file for the simulation
+    out_file = argoset.ParticleFile(
+        name=out_file_name,
+        outputdt=timedelta(minutes=5),
+        chunks=(1, 500),
+    )
+
+    # get time when the fieldset ends
+    fieldset_endtime = fieldset.time_origin.fulltime(fieldset.U.grid.time_full[-1])
```

Comment:
We allowed 6 max weeks of floating time, 3 week ship time + 3 weeks after cruise ended, but since they are deployed at different times end times would also have been different if set at 6 weeks, this was our workaround